### DISCUSSION TITLE
Publish setup size is too big.

### DISCUSSION THREAD

**Discussion by VikasSaxenaxcdify** [microsoft/WindowsAppSDK#2515] [Q&A]
How we can create a .exe setup for the winui application. When we publish the app using "create a web app" option in "package and publish" menu. Its size is too big. Is there any option available for downloading a framework from website?
I want the setup size within 10mb.

**Comment by DrusTheAxe** [comment]
>When we publish the app using "create a web app" option in "package and publish" menu

Would you be making a ClickOnce application?

>Is there any option available for downloading a framework from website?

Yes!

You don't need to use WindowsAppRuntimeInstall.exe to install the WinAppSDK runtime as long as you accomplish the same net results. This is documented as "Option 2: Install the packages directly" under the [Deploy Windows App SDK runtime](https://docs.microsoft.com/windows/apps/windows-app-sdk/deploy-unpackaged-apps#deploy-windows-app-sdk-runtime).

We're working on samples demonstrating this but they're not available yet. Here's a high level description of what the installer does:

```
FOREACH p IN *.msix
    packagemanager.AddPackageAsync(p)
    IF !p.IsFramework and IsElevated()
        packagemanager.ProvisionPackagesForAllUsersAsync(p.packagefamilyname)
BootstrapInitialize()
h = LoadLibrary("Microsoft.WindowsAppRuntime.dll")
f = GetProcAddress("MsixInstallLicenses")
f()
IF Force option
    RestartPushNotificationsLRP()
```

That's an 'earlier comment](https://github.com/microsoft/WindowsAppSDK/discussions/2099#discussioncomment-2392289) circa March'2022, plus if necessary restarting PushNotifications long-running process

[WindowsAppRuntimeInstall.exe's code is in the repository](https://github.com/microsoft/WindowsAppSDK/tree/main/installer/dev) (under `installer\dev`), for your reference. You can find `RestartPushNotificationsLRP()` in there plus other tidbits you may find helpful.

If you put the packages somewhere you can reach then you can pass their URL to `PackageManager` APIs to install them e.g. `pkgmgr.AddPackageAsync(new Uri("https://my.com/foo/bar/microsoft.windowsappruntime.1.1.msix"))`. No need to have the *.msix physically local to do the install.

@angelazhangmsft @sachintaMSFT for visibility

**Reply by eugensk** [reply]
@DrusTheAxe
Does your recipe still stand? I ask because I get this behavior on a fresh computer:
MsixInstallLicenses() returns S_OK,  
pushNotificationManager.Register() returns  0xC0EA0001 (No applicable app licenses found).

My app is an unpackaged C++ app,  it uses Windows App SDK 1.5.5 for push notifications.
When installing, as LOCAL SYSTEM, it stages framework, ddlm, singleton and main packages,  x64 variants only.
When run, as current user, it first registers the packages, then, as you suggest,  it does
```
MddBootstrap::Initialize()
h = LoadLibrary("Microsoft.WindowsAppRuntime.dll")
f = GetProcAddress("MsixInstallLicenses")
f()
...
pushNotificationManager.Register();

```

and the result is this:

```
21:34:36.993 registering packages...
21:34:37.868 Microsoft.WindowsAppRuntime.1.5_5001.178.1908.0_x64__8wekyb3d8bbwe OK
21:34:39.227 Microsoft.WinAppRuntime.DDLM.5001.178.1908.0-x6_5001.178.1908.0_x64__8wekyb3d8bbwe OK
21:34:40.024 MicrosoftCorporationII.WinAppRuntime.Main.1.5_5001.178.1908.0_x64__8wekyb3d8bbwe OK
21:34:40.024 MicrosoftCorporationII.WinAppRuntime.Singleton_5001.178.1908.0_x64__8wekyb3d8bbwe OK
21:34:40.024 loading Windows App Runtime 1.5...
21:34:40.446 OK
21:34:40.446 installing licenses...
21:34:40.446 OK
...
21:34:42.446 calling PushNotificationManager.Register()
21:34:42.461 Register failed with winrt::hresult_error c0ea0001

```

If any other Store app brings along Windows App SDK 1.5, or if I run windowsappruntimeinstall-x64.exe,
then PushNotificationManager.Register() succeeds. May be there is a sample how to install the licenses in my case, but I could not find it. The installer can't be the one, it passes licenses stored as resources to some internal implementation. Also there is
https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/win32/windowsappruntime-licensing/nf-windowsappruntime-licensing-windowsappruntime_installlicenses
but I have no idea how to use this API.  May be I can somehow call InstallLicenseFromFile() myself, or something like that? Please, tell how the licensing API is to be used.

**Reply by eugensk** [reply]
@DrusTheAxe
I looked in debugger what MsixInstallLicenses() does, and it is no-op, it returns S_OK, and that's it.
On the other hand, stock Windows App runtime installer, windowsappruntimeinstall-x64.exe, makes calls into clipc.dll. 
So, how do I install the license without bringing  in the entire stock installer?

**Reply by DrusTheAxe** [reply]
c0ea0001 = STATUS_NO_APPLICABLE_APP_LICENSES_FOUND

It's a NOOP...in the GitHub repo. The build pipeline replaces the NOOP implementation with a closed source implementation to do the actual work. This NOOP'ing is also needed for dev inner-loop as the license data usually isn't available in local dev builds.

But..I think `MsixInstallLicenses()` is bugged. I've passed along word to the dev team
https://task.ms/54266381

Thanks for reporting this!

**Reply by eugensk** [reply]
@DrusTheAxe
> It's a NOOP...in the GitHub repo. The build pipeline replaces the NOOP implementation with...

Well, then it seems the build pipeline fails to do that lately.  I have several Windows App SDKs installed from the redistributables, and I checked the version I am using, 1.5.5, and the latest, 1.6.1:

```
dumpbin.exe /exports /disasm /headers "C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.5_5001.178.1908.0_x64__8wekyb3d8bbwe\Microsoft.WindowsAppRuntime.dll" 

dumpbin.exe /exports /disasm /headers "C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.6_6000.266.2241.0_x86__8wekyb3d8bbwe\Microsoft.WindowsAppRuntime.dll" 

dumpbin.exe /exports /disasm /headers "C:\Program Files\WindowsApps\Microsoft.WindowsAppRuntime.1.6_6000.266.2241.0_x64__8wekyb3d8bbwe\Microsoft.WindowsAppRuntime.dll"
```
there are empty MsixInstallLicenses() and WindowsAppRuntime_EnsureIsLoaded():

```
    ordinal hint RVA      name
         16    F 00006D30 MsixInstallLicenses
         17   10 00006D30 WindowsAppRuntime_EnsureIsLoaded

  0000000180006D30: 33 C0              xor         eax,eax
  0000000180006D32: C3                   ret

```

**Reply by eugensk** [reply]
@DrusTheAxe

Still no-op MsixInstallLicenses() in the latest version, 
Microsoft.WindowsAppRuntime.1.7_7000.522.1444.0_x64__8wekyb3d8bbwe/Microsoft.WindowsAppRuntime.dll:

```
  000000018008A3A0: 48 83 EC 28        sub         rsp,28h
  000000018008A3A4: 33 C0              xor         eax,eax
  000000018008A3A6: EB 04              jmp         000000018008A3AC
  000000018008A3A8: 8B 44 24 30        mov         eax,dword ptr [rsp+30h]
  000000018008A3AC: 48 83 C4 28        add         rsp,28h
  000000018008A3B0: C3                 ret

       180000000 image base (0000000180000000 to 00000001801E5FFF)

         17    F 0008A3A0 MsixInstallLicenses

```
